cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include "./utilities.hpp"
+#include "./utilities.cuh"
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+/**
+ * @brief Returns a new strings column using the specified indices to select
+ * elements from the `strings` column.
+ *
+ * ```
+ * s1 = ["a", "b", "c", "d", "e", "f"]
+ * map = [0, 2]
+ * s2 = gather( s1, map )
+ * s2 is ["a", "c"]
+ * ```
+ *
+ * @param strings Strings instance for this operation.
+ * @param begin Start of index iterator.
+ * @param end End of index iterator.
+ * @param ignore_out_of_bounds If true, indices outside the column's range are ignored.
+ * @param mr Resource for allocating device memory.
+ * @param stream CUDA stream to use kernels in this method.
+ * @return New strings column containing the gather strings only
+ */
+template<typename MapIterator>
+std::unique_ptr<cudf::column> gather( strings_column_view const& strings,
+                                      MapIterator begin, MapIterator end,
+                                      bool ignore_out_of_bounds,
+                                      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                      cudaStream_t stream=0 )
+{
+    auto strings_count = std::distance(begin, end);
+    if( strings_count == 0 )
+        return std::make_unique<column>( data_type{STRING}, 0,
+                                         rmm::device_buffer{0,stream,mr},
+                                         rmm::device_buffer{0,stream,mr}, 0 );
+    auto execpol = rmm::exec_policy(stream);
+    auto strings_column = column_device_view::create(strings.parent(),stream);
+    auto d_strings = *strings_column;
+
+    // create null mask
+    rmm::device_buffer null_mask;
+    auto valid_mask = cudf::experimental::detail::valid_if( begin, end,
```

Comment:
The null mask should be uninitialized. The `cudf::groupby` logic will take care of filling in the mask.